core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLReaderBatchSource.java

```diff
@@ -0,0 +1,305 @@
+
+    StructuredRecord output = StructuredRecord.builder(DEFAULT_XML_SCHEMA)
+      .set("offset", input.getKey().get())
+      .set("filename", fileName)
+      .set("record", record)
+      .build();
+    emitter.emit(output);
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSourceContext context) {
+    super.onRunFinish(succeeded, context);
+    if (StringUtils.isNotEmpty(this.config.tableName)) {
+      try {
+        //read file tracking information updated by XMLRecordReader and put it in dataset.
+        File file = new File(this.config.tableName);
+        BufferedReader reader = new BufferedReader(new FileReader(file));
+        String line = null;
+        while ((line = reader.readLine()) != null) {
+          long processingTime = new Date().getTime();
+          processedFileTrackingTable.write(line.getBytes(Charsets.UTF_8),
+                                           String.valueOf(processingTime).getBytes(Charsets.UTF_8));
+        }
+        reader.close();
+        file.delete();
+      } catch (IOException exception) {
+        exception.printStackTrace();
+      }
+    }
+  }
+
+  /**
+   * Config class that contains all the properties needed for the XML Reader.
+   */
+  public static class XMLReaderConfig extends ReferencePluginConfig {
+    @Description(PATH_DESCRIPTION)
+    public String path;
+
+    @Nullable
+    @Description(PATTERN_DESCRIPTION)
+    public String pattern;
+
+    @Description(NODE_PATH_DESCRIPTION)
+    public String nodePath;
+
+    @Nullable
+    @Description(AFTER_PROCESS_ACTION_DESCRIPTION)
+    public String actionAfterProcess;
+
+    @Nullable
+    @Description(TARGET_FOLDER_DESCRIPTION)
+    public String targetFolder;
+
+    @Description(FILES_TO_REPROCESS_DESCRIPTION)
+    public String reprocessingReq;
+
+    @Description(TABLE_NAME_DESCRIPTION)
+    public String tableName;
+
+    @VisibleForTesting
+    public XMLReaderConfig(String referenceName, String path, @Nullable String pattern,
+                           @Nullable String nodePath, @Nullable String actionAfterProcess,
+                           @Nullable String targetFolder, String reprocessingReq, @Nullable String tableName) {
+      super(referenceName);
+      this.path = path;
+      this.pattern = pattern;
+      this.nodePath = nodePath;
+      this.actionAfterProcess = actionAfterProcess;
+      this.targetFolder = targetFolder;
+      this.reprocessingReq = reprocessingReq;
+      this.tableName = tableName;
+    }
+
+    @VisibleForTesting
+    public String getTableName() {
+      return this.tableName;
+    }
+
+    @VisibleForTesting
+    public String getReprocessingReq() {
+      return this.reprocessingReq;
+    }
+
+    @VisibleForTesting
+    public String getPath() {
+      return this.path;
+    }
+
+    @VisibleForTesting
+    public String getNodePath() {
+      return this.nodePath;
+    }
+
+    public void validateConfig() {
+      if (Strings.isNullOrEmpty(this.path)) {
```

Comment:
Yes, else does not make sense, I will remove it. To simplify means, shall I check all required fields at once and throw single message explaining missing fields? please confirm if I understood it correctly.